src/helper_functions.cpp

```diff
@@ -17,15 +17,11 @@ std::string string_from_unterminated(const char* data, size_t data_length) {
 
 namespace {
 std::string utf16ToUtf8(const std::u16string& wstr) {
-#ifdef _MSC_VER
-#pragma warning(push)
-#pragma warning(disable : 4244)
-  auto str = std::string(wstr.begin(), wstr.end());
-#pragma warning(pop)
-#else
-  auto str = std::string(wstr.begin(), wstr.end());
-#endif
-  str.erase(std::remove(str.begin(), str.end(), '\0'), str.end());
+  std::string str;
+  str.reserve(wstr.size());
+  for (auto w : wstr)
+    if (w != '\0')
+      str.push_back(static_cast<char>(w));
```

Comment:
This doesn't look like proper UTF-16 to UTF-8 conversion. Isn't there some standard library function that we could use?

Reply:
No.

I tried using iconv. Didn't work either.

codecvt is deprecated in C++17 and removed in C++20. Probably because it does the same as this code.

Reply:
ping @mohamedchebbii

Reply:
Oops, it looks like [we actually might have to implement it,](https://stackoverflow.com/a/69410299) but it also [looks like we have already?](https://github.com/Exiv2/exiv2/blob/3403dabec247a41206138645b5045fd2cfddf2a3/src/convert.cpp#L1432-L1515)

Reply:
Tried that. Same result.

Reply:
So, revert https://github.com/Exiv2/exiv2/commit/0fa22ed55d1088eacc5ef063b7307fa1b4d031e6 for now?

Reply:
No. codecvt is also broken.

Reply:
Hello,
Yes indeed, It seems that  std::codecvt_utf8_utf16 facet is deprecated in C++17 and will be removed in future versions of the language. we can use the std::wstring_convert class with the std::codecvt_utf8<char16_t> facet instead:

```
#include <locale>
#include <codecvt>
#include <string>
#include<iostream>

int main() {
    std::u16string u16str = u"текст";
    std::wstring_convert<std::codecvt_utf8_utf16<char16_t>, char16_t> convert;
    std::string str = convert.to_bytes(u16str);
    std::cout<<str;
    return 0;
}
```

what do you think @neheb  ?

Reply:
I think that's deprecated too. In any case, I can't find a solution where exiv2 -P X shows the unicode result properly.

Reply:
Shall we open a separate issue about it?

Reply:
Found a fix.